src/callouts.rs

```diff
@@ -106,6 +196,235 @@ impl Display for State {
+                self.path.clone(),
+                event
+            );
+            return Err(CalloutError::EventNotSupported(self.path.clone(), event));
+        }
+        Ok(())
+    }
+}
+
+impl AsRef<Path> for CalloutScript {
+    fn as_ref(&self) -> &Path {
+        &self.path
+    }
+}
+
+pub struct CalloutScripts {
+    callouts: Vec<CalloutScript>,
+}
+
+static CALLOUT_SCRIPTS: Mutex<CalloutScripts> = Mutex::new(CalloutScripts::new());
+
+fn find_callout_script(dev: &mut MDev) -> Result<CalloutScript, CalloutError> {
+    return CALLOUT_SCRIPTS.lock().unwrap().find_script(dev);
+}
+
+impl CalloutScripts {
+    pub const fn new() -> Self {
+        CalloutScripts {
+            callouts: Vec::new(),
+        }
+    }
+
+    fn find_script(&mut self, dev: &mut MDev) -> Result<CalloutScript, CalloutError> {
+        // check already found scripts
+        let mdev_type = dev.mdev_type().expect("mdev_type is required on device");
+        debug!("Looking up callout script for mdev type '{:?}'", mdev_type);
+        for cs in &self.callouts {
+            if cs.mdev_type.eq_ignore_ascii_case(mdev_type) {
+                debug!(
+                    " Looked up callout script for mdev type '{:?}': {:?}",
+                    mdev_type, cs.path
+                );
+                return Ok(cs.clone());
+            }
+        }
+        debug!(
+            "Lookup failed starting to search for mdev type '{:?}'",
+            mdev_type
+        );
+        // search directories for mdev_type parent tuple
+        for dir in dev.env.callout_dirs() {
+            debug!(
+                "Searching in directory {:?} for mdev type '{:?}'",
+                mdev_type, dir
+            );
+            match self.callout_dir_search(&mut dev.clone(), dir.clone()) {
+                Ok(cs) => {
+                    debug!(
+                        " Found callout script for mdev type '{:?}': {:?}",
+                        mdev_type, cs.path
+                    );
+                    self.callouts.push(cs.clone());
+                    return Ok(cs);
+                }
+                Err(CalloutError::NoMatchingScript) => {
+                    debug!(" Search returned without match... continue");
+                    continue;
+                }
+                Err(e) => {
+                    debug!(" Search returned with error {:?}", e);
+                    return Err(e);
+                }
+            }
+        }
+        // at this point no script was found
+        debug!(
+            "Searching callout script for mdev type '{:?}' ended without result",
+            mdev_type
+        );
+        Err(CalloutError::NoMatchingScript)
+    }
+
+    fn callout_dir_search(
+        &mut self,
+        dev: &mut MDev,
+        dir: PathBuf,
+    ) -> Result<CalloutScript, CalloutError> {
+        if !dir.is_dir() {
+            return Err(CalloutError::NoMatchingScript);
+        }
```

Comment:
To avoid cloning above:

```suggestion
    fn callout_dir_search<P: AsRef<Path> + std::fmt::Debug>(
        &mut self,
        dev: &MDev,
        dir: P,
    ) -> Result<CalloutScript, CalloutError> {
        if !dir.as_ref().is_dir() {
            return Err(CalloutError::NoMatchingScript);
        }
```

Reply:
This method has been eliminated when rebasing on your refactoring.